modules/apm/src/main/java/org/elasticsearch/telemetry/apm/internal/metrics/AbstractInstrument.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.telemetry.apm.internal.metrics;
+
+import io.opentelemetry.api.metrics.Meter;
+
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.telemetry.metric.Instrument;
+import org.elasticsearch.telemetry.metric.MetricName;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class AbstractInstrument<T, I> implements Instrument {
+    private final AtomicReference<I> instrumentRef;
+    private final MetricName name;
+    private final String description;
+    private final T unit;
+
+    public AbstractInstrument(Meter meter, MetricName name, String description, T unit) {
+        this.instrumentRef = new AtomicReference<>(buildInstrument(meter));
```

Comment:
there is a problem with this approach.
1. buildInstrument implementations are relying on the `getName`, getDescription and getUnit methods. THose are not initialised yet.
2. the implementing classes might implement this 'in a wrong way' (unlikely but still) leading to concurrency problems. Perhaps we can mitigate this by sealed class.

I think we should change the signature of buildInstrument method to accept name, description and unit and we could pass it as a constructor parameter

Reply:
Discussed offline, moved buildInstrument last to avoid other fields being uninitialized.